### ISSUE TITLE
font lies

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [arkenfox/TZP#28]
here is the current fpjs2 style font enumeration

```js
function get_fpjs2() {
	/* based on https://github.com/Valve/fingerprintjs2 */
	return new Promise(resolve => {

		// vars
		let baseFonts = ['monospace','sans-serif','serif'],
			t0 = performance.now()
		// elements
		let h = document.getElementsByTagName('body')[0]
		let baseFontsDiv = document.createElement('div')
		let fontsDiv = document.createElement('div')
		let defaultWidth = {}
		let defaultHeight = {}
		let createSpan = function() {
			let s = document.createElement('spanFP')
			s.style.position = "absolute"
			s.style.left = "-9999px"
			s.style.fontSize = "256px"
			s.style.fontStyle = "normal"
			s.style.fontWeight = "normal"
			s.style.letterSpacing = "normal"
			s.style.lineBreak = "auto"
			s.style.lineHeight = "normal"
			s.style.textTransform = "none"
			s.style.textAlign = "left"
			s.style.textDecoration = "none"
			s.style.textShadow = "none"
			s.style.whiteSpace = "normal"
			s.style.wordBreak = "normal"
			s.style.wordSpacing = "normal"
			s.innerHTML = fntStrA
			return s
		}
		// creates a span and load the font to detect and a base font for fallback
		let createSpanWithFonts = function(fontToDetect, baseFont) {
			let s = createSpan()
			s.style.fontFamily = "'" + fontToDetect + "'," + baseFont
			return s
		}
		// creates spans for the base fonts and adds them to baseFontsDiv
		let initializeBaseFontsSpans = function() {
			let spans = []
			for (let index=0, length=baseFonts.length; index<length; index++) {
				let s = createSpan()
				s.style.fontFamily = baseFonts[index]
				baseFontsDiv.appendChild(s)
				spans.push(s)
			}
			return spans
		}
		// creates spans for the fonts to detect and adds them to fontsDiv
		let initializeFontsSpans = function() {
			let spans = {}
			for (let i = 0; i < fntList.length; i++) {
				let fontSpans = []
				for (let j=0, numDefaultFonts = baseFonts.length; j< numDefaultFonts; j++) {
					let s = createSpanWithFonts(fntList[i], baseFonts[j])
					fontsDiv.appendChild(s)
					fontSpans.push(s)
				}
				spans[fntList[i]] = fontSpans // Stores {fontName : [spans for that font]}
			}
			return spans
		}
		// compare
		let present = function(fontSpans) {
			let r = false
			for (let i=0; i < baseFonts.length; i++) {
				r = (fontSpans[i].offsetWidth !== defaultWidth[baseFonts[i]] || fontSpans[i].offsetHeight !== defaultHeight[baseFonts[i]])
				if (r) {return r}
			}
			return r
		}
		// stuff
		let baseFontsSpans = initializeBaseFontsSpans()
		h.appendChild(baseFontsDiv)
		for (let index=0, length = baseFonts.length; index<length; index++) {
			defaultWidth[baseFonts[index]] = baseFontsSpans[index].offsetWidth // width for the default font
			defaultHeight[baseFonts[index]] = baseFontsSpans[index].offsetHeight // height for the default font
		}
		let fontsSpans = initializeFontsSpans()
		h.appendChild(fontsDiv)
		// detect
		let found = []
		for (let i=0; i < fntList.length; i++) {
			if (present(fontsSpans[fntList[i]])) {found.push(fntList[i])}
		}
		// cleanup
		h.removeChild(fontsDiv)
		h.removeChild(baseFontsDiv)

		// resolve
		return resolve("fonts:" + hash)
	})
}
```

here is a fiddle from abraham : https://jsfiddle.net/j6gc8amy/10/

```html
<span id="show-font"></span>
```

```css
#show-font {
    --font: '';
	position: absolute !important;
    font-size: 256px !important;
    font-style: normal !important;
    font-weight: normal !important;
    letter-spacing: normal !important;
    line-break: auto !important;
    line-height: normal !important;
    text-transform: none !important;
    text-align: left !important;
    text-decoration: none !important;
    text-shadow: none !important;
    white-space: normal !important;
    word-break: normal !important;
    word-spacing: normal !important;
}

#show-font::after {
    font-family: var(--font);
    content: 'mmmmmmmmmmlli';
}
```

```js
const span = document.getElementById('show-font')
const baseFonts = ['monospace', 'sans-serif', 'serif']
const base = {}
baseFonts.forEach(font => {
	span.style.setProperty('--font', font)
	const { width, height } = span.getClientRects()[0]
	base[font] = { width, height }
	return	
})

const families = fontList.reduce((acc, font) => {
	baseFonts.forEach(baseFont => acc.push(`'${font}', ${baseFont}`))
	return acc
},[])
const detected = new Set()


families.forEach(family => {
    span.style.setProperty('--font', family)
	const basefont = /, (.+)/i.exec(family)[1]
    const { width, height } = span.getClientRects()[0]
	const supported = (
		width != base[basefont].width || height != base[basefont].height
	)
	if (supported) {
		const font = /\'(.+)\'/i.exec(family)
		return detected.add(font[1])
	}
	return
})
console.log(`${detected.size}/${fontList.length}`, [...detected].join('\n'))

})()
```

also see https://github.com/abrahamjuliot/creepjs/issues/100

I tested this with chameleon's font protection: TZP reports 0 fonts for `fpjs2` and `font fallback` methods. Tested fiddle with and without chameleon = same result: i.e it found the fonts

more to follow

**Comment by Thorin-Oakenpants** [comment]
questions, thoughts:
- lets say I use this for TZP's `fpjs2` method, can I also use it in `font fallback` method?
- perf: won't know until we test it
   - current tzp is approx 120ms (flip logPerf global var to true and do section reruns) - that's using the windows list of 486 fonts
- creep method could do with returning a lie attempt boolean: i'm sure we can work that in
- current TZP uses a create span function which sets the style each time: I remember reading about this being a necessary step due to scripts etc changing it
- creep is re-using a single span, tzp creates one per font in the list
- I thought the string I was using had appended unicode chars, but right now it's just using `mmmLLLmmmWWWwwwmmmllliii`
  - probably easiest to set it from JS rather than css

is there something I'm missing? Lets see if I can't get this to work

**Comment by Thorin-Oakenpants** [comment]
i don't get it: looking at your patches I don't see how I am supposed to add `--font` to style

```js
function get_fpjs2() {
	// expecting 3710e8c986c3cff7366afc10f000654d604093e0 [150/486]
	return new Promise(resolve => {
		if (isFF) {
			// vars
			let baseFonts = ['monospace','sans-serif','serif'],
				t0 = performance.now()

			// create span
			let span = document.createElement('show-font')
			// apply style
			span.style.position = "absolute"
			span.style.left = "-9999px"
			span.style.fontSize = "256px"
			span.style.fontStyle = "normal"
			span.style.fontWeight = "normal"
			span.style.letterSpacing = "normal"
			span.style.lineBreak = "auto"
			span.style.lineHeight = "normal"
			span.style.textTransform = "none"
			span.style.textAlign = "left"
			span.style.textDecoration = "none"
			span.style.textShadow = "none"
			span.style.whiteSpace = "normal"
			span.style.wordBreak = "normal"
			span.style.wordSpacing = "normal"
			// apply text
			span.innerHTML = fntStrA

			// now what

			// perf
			if (logPerf) {debug_log("fpjs2 [fonts]",t0)}

			// temp
			let hash = "coming"

			return resolve("fonts:" + hash)
		} else {
			// non-FF
			return resolve("fonts:" + zNA)
		}
	})
}
```

**Comment by Thorin-Oakenpants** [comment]
I'm just about ready to give up, this is so depressing. Two years pissing around with a deep knowledge of Firefox fingerprinting (and FPing in general) and a vision to move it forward, **_but lacking in JS/CSS know how_** <-- this is my bottleneck. Meanwhile, others (and good on them) can just whip things up in minutes, and build more robust solutions than my shitty code (and use things like classes and prototypes and whizzy do stuff and decent error handling) and other projects like creep (and good on it) are up and running in months, adding things I'd already thought of and tried to do (or are on my list) like all windows properties (which became iframe properties). I still have a really big one no-one has relaly thought of or exploited, but I'm not sure how to tackle it.

I'm wasting my time. Imagine if a dozen people from Mozilla, or mozillians with the right skills, had helped out in their spare time: this would have been finished and on its own domain 18 months ago - and the data reporting / analysis section would have been done as well. The world is full of coders and only two people ever really helped, and I hate badgering them

Seriously .. where can I recruit some people? Because at this rate, I'm about to off myself out of depression and frustration

**Comment by abrahamjuliot** [comment]
Hmmm...🤔 I'm not sure why _Chameleon_ does not spoof this? _Font Fingerprint Defender_ effectively adds integer noise to `offsetWidth` and `offsetHeight`, but this can be detected and bypassed with `scrollWidth`, `clientWidth`, `DomRect`, etc... I think even CSS media queries + getComputedStyle can test and detect the element's width and height. From what I have observed, the main difference from `fpjs2` is I use `span.style.setProperty()` instead of `s.style.fontFamily`.  The use of one span keeps it tidy, but I'm finding there is no major performance gain (if any), and the css psuedo content + css variable approach only make it harder to detect (since I'm not accessing the `font-family` or `fontFamily` property directly). 

The pseudo content/variable style can be imported via a style tag with `innerHTML` or  this approach may be discarded and the `font-family` can be set directly with `span.style.setProperty('font-family', family)` -- in this case the span should contain the characters to test. In short, I had hopes this would improve performance, but I don't think that is the case.

```js
element.innerHTML = `
<style>
#${id}-detector {
	--font: '';
	position: absolute !important;
	left: -9999px!important;
	font-size: 256px !important;
	font-style: normal !important;
	font-weight: normal !important;
	letter-spacing: normal !important;
	line-break: auto !important;
	line-height: normal !important;
	text-transform: none !important;
	text-align: left !important;
	text-decoration: none !important;
	text-shadow: none !important;
	white-space: normal !important;
	word-break: normal !important;
	word-spacing: normal !important;
	/* in order to test scrollWidth, clientWidth, etc. */
	padding: 0 !important;
	margin: 0 !important;
}
#${id}-detector::after {
	font-family: var(--font);
	content: 'mmmmmmmmmmlli';
}
</style>
<span id="${id}-detector"></span>
`
// set font with  with span.style.setProperty('--font', font)
```
Or
```js
element.innerHTML = `
<style>
#${id}-detector {
	position: absolute !important;
	left: -9999px!important;
	font-size: 256px !important;
	font-style: normal !important;
	font-weight: normal !important;
	letter-spacing: normal !important;
	line-break: auto !important;
	line-height: normal !important;
	text-transform: none !important;
	text-align: left !important;
	text-decoration: none !important;
	text-shadow: none !important;
	white-space: normal !important;
	word-break: normal !important;
	word-spacing: normal !important;
	/* in order to test scrollWidth, clientWidth, etc. */
	padding: 0 !important;
	margin: 0 !important;
}
</style>
<span id="${id}-detector">mmmmmmmmmmlli</span>
`
// set font with span.style.setProperty('font-family', family)
```

Here are 2 concept versions to test:
- 64 font test (final with lie detection): https://jsfiddle.net/8p13cj5o/ (35-70ms, ~350ms on load)
- 485 font test w/o lie detection: https://jsfiddle.net/djgwycbp/ (160-250ms, ~1000ms on load)

**Comment by Thorin-Oakenpants** [comment]
no sure about the perf either. my current tzp method (multiple spans, no lie detection) is about 120ms

for example. I changed the two tests above to use my windows 486 font list
- the ex-64 test = 100ms <-- I don't see anything about lie detection here: I just get a single array of fonts returned
- the ex-485 test = 160-180ms

So I think there are perf gains by using a single span, but the lie detection extras pulls that back

On spoofing etc, I need a working model first before I can see what works and what doesn't. I had already thought of randomness of offset. Chameleon I think blocks rather than spoofs. I was also thinking of other ways to get fonts measurements: such as domrect, textmetrics etc - but they could be spoofed as well.

I'll see if I can work this out again without getting depressed. I'm just trying to follow your examples

**Comment by Thorin-Oakenpants** [comment]
here's where I'm at: perf is way slower: I seem to have a promise within a promise: I don't know how to fix that and that's not what is making it slower

![fonts](https://user-images.githubusercontent.com/16656956/102167497-6bcbf300-3e86-11eb-911e-5110c80e39cd.png)

so I have:
- three `font` fields to fill in with a hash + counts : e.g `[150/486]` : the fntList is a global var so I can get that count no problems
- decide which result of the three to use for the section hash (if lies = true) and output the display of the font names found
   - I would assume if lies is true: then the array with the most items is the correct one?
   - can all three be lies?
- easy to do once I can parse, but I'll add a 22nd metric on attempted font lies (so as to not pollute the actual font hash)

What I'm not sure how to do is parse that object for the arrays, or where is best to do it: e.g. I could do the steps above before returning the results from getFonts - but I can't work out how to deal with them: e.g. detectedViaScroll is a `Set` not an array

Some basic parsing help here would be welcome

PS: I added code to remove the font-fingerprint object

---

code (I wish github used smaller indent tabs)
```js
function get_fpjs2() {
	/* https://github.com/abrahamjuliot/creepjs */
	return new Promise(resolve => {
		if (isFF) {
			const getFonts = () => {
				return new Promise(resolve => {
					try {
						const createLieDetector = () => {
							let invalidDimensions = []
							return {
								getInvalidDimensions: () => invalidDimensions,
								compute: ({
									scrollWidth, scrollHeight, offsetWidth, offsetHeight, clientWidth, clientHeight
								}) => {
									const invalid = (
										scrollWidth != offsetWidth ||
										scrollWidth != clientWidth ||
										scrollHeight != offsetHeight ||
										scrollHeight != clientHeight
									)
									if (invalid) {
										invalidDimensions.push({
											width: [scrollWidth, offsetWidth, clientWidth],
											height: [scrollHeight, offsetHeight, clientHeight]
										})
									}
									return
								}
							}
						}

						const detectLies = createLieDetector()
						const doc = document // or iframe.contentWindow.document
						const id = `font-fingerprint`
						const div = doc.createElement('div')
						div.setAttribute('id', id)
						doc.body.appendChild(div)
						doc.getElementById(id).innerHTML = `
							<style>
								#${id}-detector {
									--font: '';
									position: absolute !important;
									left: -9999px!important;
									font-size: 256px !important;
									font-style: normal !important;
									font-weight: normal !important;
									letter-spacing: normal !important;
									line-break: auto !important;
									line-height: normal !important;
									text-transform: none !important;
									text-align: left !important;
									text-decoration: none !important;
									text-shadow: none !important;
									white-space: normal !important;
									word-break: normal !important;
									word-spacing: normal !important;
									/* in order to test scrollWidth, clientWidth, etc. */
									padding: 0 !important;
									margin: 0 !important;
								}
								#${id}-detector::after {
									font-family: var(--font);
									content: '` + fntStrA + `';
								}
								</style>
							<span id="${id}-detector"></span>`

						const span = doc.getElementById(`${id}-detector`)
						const detectedViaScroll = new Set()
						const detectedViaOffset = new Set()
						const detectedViaClient = new Set()
						const baseFonts = ['monospace', 'sans-serif', 'serif']
						const base = baseFonts.reduce((acc, font) => {
							span.style.setProperty('--font', font)
							const dimensions = {
								scrollWidth: span.scrollWidth,
								scrollHeight: span.scrollHeight,
								offsetWidth: span.offsetWidth,
								offsetHeight: span.offsetHeight,
								clientWidth: span.clientWidth,
								clientHeight: span.clientHeight
							}
							detectLies.compute(dimensions)
							acc[font] = dimensions
									return acc
						}, {})

						const families = fntList.reduce((acc, font) => {
							baseFonts.forEach(baseFont => acc.push(`'${font}', ${baseFont}`))
							return acc
						}, [])

						families.forEach(family => {
							span.style.setProperty('--font', family)
							const basefont = /, (.+)/.exec(family)[1]
							const dimensions = {
								scrollWidth: span.scrollWidth,
								scrollHeight: span.scrollHeight,
								offsetWidth: span.offsetWidth,
								offsetHeight: span.offsetHeight,
								clientWidth: span.clientWidth,
								clientHeight: span.clientHeight
							}
							detectLies.compute(dimensions)
							const font = /\'(.+)\'/i.exec(family)[1]
							if (dimensions.scrollWidth != base[basefont].scrollWidth ||
								dimensions.scrollHeight != base[basefont].scrollHeight) {
								detectedViaScroll.add(font)
							}
							if (dimensions.offsetWidth != base[basefont].offsetWidth ||
								dimensions.offsetHeight != base[basefont].offsetHeight) {
								detectedViaOffset.add(font)
							}
							if (dimensions.clientWidth != base[basefont].clientWidth ||
								dimensions.clientHeight != base[basefont].clientHeight) {
								detectedViaClient.add(font)
							}
							return
						})
						const fontsScroll = [...detectedViaScroll]
						const fontsOffset = [...detectedViaOffset]
						const fontsClient = [...detectedViaClient]
						return resolve({
							lies: !!detectLies.getInvalidDimensions().length, fontsScroll, fontsOffset, fontsClient
						})
					} catch (error) {
						console.error(error)
						return resolve("fonts:error")
					}
				})
			}

			// run it
			let t0 = performance.now()
			getFonts().then(res => {
				// cleanup
				try {document.getElementById("font-fingerprint").remove()} catch(e) {}
				// output
					// hash each of the three results
					// decide which one to use for display results + returned hash
					// handle "fonts:error"
				console.log(res)


				//let hash = sha1(found.join())
				//dom.fontFPJS2label = hash
				//dom.fontFPJS2Found.innerHTML = (found.length > 0 ? found.join(", ") : "no fonts detected")
				//dom.fontFPJS2.innerHTML = hash + s12 + "["+found.length+"/"+fntList.length+"]" + sc

				// unhide
				dom.fontFPJS2Found.style.color = zshow
				// cleanup details
				if (stateFNT == true) {showhide("table-row","F1","&#9650; hide")}
				// perf
				if (logPerf) {debug_log("creepy fpjs2 [fonts]",t0)}
				debug_log("creepy fpjs2 [fonts]",t0) // temp
				// resolve
				let hash = "coming"
				return resolve("fonts:" + hash)
			})
		} else {
			// non-FF
			return resolve("fonts:" + zNA)
		}
	})
}
```

**Comment by abrahamjuliot** [comment]
> promise in a promise

Since these are stand alone functions, I recommend placing `getFonts` outside of `get_fpjs2`, and we could also move `createLieDetector` outside of `getFonts`. And, `getFonts` can take `fntList` and `fntStrA` as arguments. Example: https://jsfiddle.net/rkoy8p5c

> parsing the response

The response can be parsed using destructuring or assignment. The `Sets` are cast to arrays in the resolve using `[...set]`, so the promise response will yield arrays.
```js
const { lies, fontsScroll, fontsOffset, fontsClient } = res
```
or
```js
const lies = res.lies 
const fontsScroll = res.fontsScroll 
const fontsOffset = res.fontsOffset 
const fontsClient = res.fontsClient 
```

> lies

I'm not sure any of the 3 lists can be trusted if lies are found. Lies can yield a longer or shorter lists and then go on to poison 1, 2, or all 3 of the lists at random. But, we could give greater trust to the 2 lists that match. It would be difficult to mirror the exact poison pattern on 2 lists. My current response to detecting this lie is to distrust all 3 lists and value `lied: true` as stable entropy (the fingerprint and extent of the lie is kept in loose), but I'm thinking I should trust to the 2 that match (if they exist). What do you think? Maybe there is a better approach to this.

Also, I display scroll in the template view, but I might add a css media query 4th list and give it 100% trust it if it's not blocked.

**Comment by Thorin-Oakenpants** [comment]
might take a break ... I'm getting there .. I used Array.from so I could handle stuff better

I'm outputting the display within getFonts() - but it's almost as if I need a lies per Offset, Client, Scroll

```js
	// display
	let foundO = Array.from(detectedViaOffset)
	let hashO = sha1(foundO.join())
	dom.fontOffset.innerHTML = hashO + s12 + "["+foundO.length+"/"+fntList.length+"]" + sc
	let foundC = Array.from(detectedViaClient)
	let hashC = sha1(foundC.join())
	dom.fontClient.innerHTML = hashC + s12 + "["+foundC.length+"/"+fntList.length+"]" + sc
	let foundS = Array.from(detectedViaScroll)
	let hashS = sha1(foundS.join())
	dom.fontScroll.innerHTML = hashS + s12 + "["+foundS.length+"/"+fntList.length+"]" + sc

	// lies?
	if (detectLies.getInvalidDimensions().length == 0) {
		// none
		dom.fontLabel = hashO
		dom.fontFound.innerHTML = (foundO.length > 0 ? foundO.join(", ") : "no fonts detected") 
	} else {
		// decide which one to use for display results + returned hash
	}
```

![fonts](https://user-images.githubusercontent.com/16656956/102181147-f706b200-3ea1-11eb-97ac-c70dc81eda8b.png)

^^ note the creepJS credits

I'm so at sea over all these promises and the way they can be written with => and ({ blah }) bits. I don't care where I get the results and calculate what to do with them. So yeah, some restructuring might help: I'm just losing my head over how

**Comment by Thorin-Oakenpants** [comment]
fp return - yes, this is my dilemma  as well

```
	// lies?
	if (detectLies.getInvalidDimensions().length == 0) {
		// none
		dom.fontLabel = hashO
		dom.fontFound.innerHTML = (foundO.length > 0 ? foundO.join(", ") : "no fonts detected")
	} else {
		// decide which one to use for display results + returned hash
	}
```

So I had thoughts
- if two match then that's most likely for realsies
- the highest must be the most accurate (if all three differ)
  - or can't trust any if lies = true
- does counting lies per method help here?

What I hope to return from for the section hash is an array of two items
- `fonts:hash` or `fonts:n/a` or `fonts:error` or `fonts:can't tell`
- `_spoofing_attempt_fonts:true` (or false)

**Comment by Thorin-Oakenpants** [comment]
hmmm .. interesting... 

I've noticed that RFP seems to make TZP slower: like quite a bit (300 to 400+ ms) - and that's on my main FF which has webGL and webAudio disabled (so even less to run). Now I think I know why - fonts. Seems to fit in with RFP protecting fonts in FF80, because while I didn't keep notes or anything, and I'm not looking to compare RFP time with non-RFP ... it kinda makes sense when I noticed it.

Using css `--font` rather than `setProperty('font-family', family)`, when RFP is on ... is way faster. RFP uses a list of kBaseFonts and kLangPackFonts, and is binary searched (from the moz dev). Note: RFP font protection is not bypassed, but I guess setProperty does a check on the font - IDK, but sounds semi-reasonable to me.

RFP ignores the pref (`layout.css.font-visibility.level`) level and uses just kBaseFonts (same as pref = `1`)
[1] https://searchfox.org/mozilla-central/search?path=StandardFonts*.inc

If this is true, then the creepy getFonts is the bees knees .. "This Is the Way" - mando

**Comment by Thorin-Oakenpants** [comment]
https://jsfiddle.net/rkoy8p5c : not seeing the missing step: returning getFonts() to get_fonts() (I renamed it from fpjs2)

PS: fntList and fntStrA are global vars already set, no need to pass them as parameters

```js
const createLieDetector = () => {
	/* https://github.com/abrahamjuliot/creepjs */
	let invalidDimensions = []
	return {
		getInvalidDimensions: () => invalidDimensions,
			compute: ({
				scrollWidth,
				scrollHeight,
				offsetWidth,
				offsetHeight,
				clientWidth,
				clientHeight
		}) => {
			const invalid = (
				scrollWidth != offsetWidth ||
				scrollWidth != clientWidth ||
				scrollHeight != offsetHeight ||
				scrollHeight != clientHeight
			)
			if (invalid) {
				invalidDimensions.push({
					width: [scrollWidth, offsetWidth, clientWidth],
					height: [scrollHeight, offsetHeight, clientHeight]
				})
			}
			return
		}
	}
}

const getFonts = () => {
	/* https://github.com/abrahamjuliot/creepjs */
	return new Promise(resolve => {
		if (!isFF) {
			return resolve("fonts:n/a")
		} else {
			try {
				const detectLies = createLieDetector()
				const doc = document // or iframe.contentWindow.document
				const id = `font-fingerprint`
				const div = doc.createElement('div')
				div.setAttribute('id', id)
				doc.body.appendChild(div)
				doc.getElementById(id).innerHTML = `
					<style>
					#${id}-detector {
						--font: '';
						position: absolute !important;
						left: -9999px!important;
						font-size: 256px !important;
						font-style: normal !important;
						font-weight: normal !important;
						letter-spacing: normal !important;
						line-break: auto !important;
						line-height: normal !important;
						text-transform: none !important;
						text-align: left !important;
						text-decoration: none !important;
						text-shadow: none !important;
						white-space: normal !important;
						word-break: normal !important;
						word-spacing: normal !important;
						/* in order to test scrollWidth, clientWidth, etc. */
						padding: 0 !important;
						margin: 0 !important;
					}
					#${id}-detector::after {
						font-family: var(--font);
						content: '` + fntStrA + `';
					}
					</style>
					<span id="${id}-detector"></span>`

				const span = doc.getElementById(`${id}-detector`)
				const detectedViaScroll = new Set()
				const detectedViaOffset = new Set()
				const detectedViaClient = new Set()
				const baseFonts = ['monospace', 'sans-serif', 'serif']
				const base = baseFonts.reduce((acc, font) => {
					span.style.setProperty('--font', font)
					const dimensions = {
						scrollWidth: span.scrollWidth,
						scrollHeight: span.scrollHeight,
						offsetWidth: span.offsetWidth,
						offsetHeight: span.offsetHeight,
						clientWidth: span.clientWidth,
						clientHeight: span.clientHeight
					}
				detectLies.compute(dimensions)
					acc[font] = dimensions
					return acc
				}, {})

				const families = fntList.reduce((acc, font) => {
					baseFonts.forEach(baseFont => acc.push(`'${font}', ${baseFont}`))
					return acc
				}, [])

				families.forEach(family => {
					span.style.setProperty('--font', family)
					const basefont = /, (.+)/.exec(family)[1]
					const dimensions = {
						scrollWidth: span.scrollWidth,
						scrollHeight: span.scrollHeight,
						offsetWidth: span.offsetWidth,
						offsetHeight: span.offsetHeight,
						clientWidth: span.clientWidth,
						clientHeight: span.clientHeight
					}
					detectLies.compute(dimensions)
					const font = /\'(.+)\'/.exec(family)[1]
					if (dimensions.scrollWidth != base[basefont].scrollWidth ||
						dimensions.scrollHeight != base[basefont].scrollHeight) {
						detectedViaScroll.add(font)
					}
					if (dimensions.offsetWidth != base[basefont].offsetWidth ||
						dimensions.offsetHeight != base[basefont].offsetHeight) {
						detectedViaOffset.add(font)
					}
					if (dimensions.clientWidth != base[basefont].clientWidth ||
						dimensions.clientHeight != base[basefont].clientHeight) {
						detectedViaClient.add(font)
					}
					return
				})
				const fontsScroll = [...detectedViaScroll]
				const fontsOffset = [...detectedViaOffset]
				const fontsClient = [...detectedViaClient]
				return resolve({
					lies: !!detectLies.getInvalidDimensions().length,
					fontsScroll,
					fontsOffset,
					fontsClient
				})
			} catch (error) {
				console.error(error)
				return resolve("fonts:error")
			}
		}
	})
}

function get_fonts() {
	/* https://github.com/abrahamjuliot/creepjs */

	//let t0 = performance.now()
	// how do I perf this?

	return new Promise(resolve => {
		// return promise to outputFonts() which is the whole fonts section being run

		//where is res ?
		//const { lies, fontsScroll, fontsOffset, fontsClient } = res

	})
}

function outputFonts() {
	Promise.all([
		get_fonts(),
		// snip
	]).then(function(results){
		// do section stuff
	})
}

```

**Comment by abrahamjuliot** [comment]
Looking good. I added this...

- in `getFonts`, removed the large else block/nest since the !isFF will exit or execute everything else by default
- get and handle a response in `get_fonts`, but you could move the computation here to inside getFonts and compute the performance once the fingerprint is finished (before extra computation or output to the dom):

https://jsfiddle.net/ro8w6v52/
```js
function get_fonts() {
    return new Promise(resolve => {
        // optional: try/catch this block
        // return promise to outputFonts() which is the whole fonts section being run
       let t0 = performance.now()
		
        // res
        getFonts().then(res => {
	    // calc final performance here
	    // const { lies, fontsScroll, fontsOffset, fontsClient } = res
	    // ...
	    // const data = ...
			
	    // return resolve(data)
	})
	.catch(error => console.error(error))
    })
}
```

**Comment by Thorin-Oakenpants** [comment]
ahead of you .... about to update TZP: just need to work out the logic for lies and which (if any) hash to return

if else block: good thinking

PS: RFP perf improvements are massive when re-running the whole thing or on page load: the function still takes around the same time: RFP rounds it out at 400ms - which is about the same as non-RFP: so I don't know what is changing. But clearly the fonts setProperty method is bottlenecking somewhere. Might try testing with moving fonts to last on the old tzp method: but it might be academic now

**Comment by Thorin-Oakenpants** [comment]
Lie logic
- if two of the three results are the same: then use one of those
   - but what if those two same methods were lies and the third one is the real one
   - also, if the outlying third result has a higher count: does that mean the spoofing failed to hide it, or does it mean the spoofing inflated the results
- if all three are different
   - can't trust anyone?

What are the extension spoofers doing: trying to hide fonts or add noise by returning everything as found?

What are your thoughts? What about this extra css media check?

**Comment by abrahamjuliot** [comment]
I'm leaning towards distrusting all 3 if there is a lie. At that point, I don't know which list if any is telling the truth. The poison can yield both high and low results at random. Font Fingerprint Defender gets the `DOMRect` width and applies 1 digit offest at random to `offestWidth` and `offsetHeight`. Cydec has a font protection. It seemed to only block, but it's been awhile since I tested it. ScriptSafe has font protection in canvas only. 

I don't think media queries will work 😭. I was thinking there's some magic I could perform to calculate empty space in the viewport (first validating the veiwport to the media query). I really need a container query which is sadly [not a thing yet](https://css-tricks.com/lets-not-forget-about-container-queries/). But, [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver/ResizeObserver) might be a an option. More research needed.

There's also 4th list "computed" I might add using `getComputedStyle`. It should match the other 3. 
```js
// computed
const pixelsToInt = pixels => Math.floor(+pixels.replace('px',''))
const style = getComputedStyle(span)
const dimensions = {
    width: pixelsToInt(style.width),
    height: pixelsToInt(style.height),
}
```

**Comment by kkapsner** [comment]
> can't seem to find any other FF font protection extensions

You know my opinion on that: https://github.com/kkapsner/CanvasBlocker/issues/200#issuecomment-410053873

**Comment by Thorin-Oakenpants** [comment]
Tor Project etc know the score. The only way to do it is to bundled all users with the exact same fonts, font versions, and font settings - but bundling all fonts is prohibitively huge. All those ilGur [Defender](https://addons.mozilla.org/en-US/firefox/user/11791768/) extensions are pretty lacking, just wanted something to test

I added simulation code in the end and tested all possible outcomes. I think there's a few more ways to get fonts: textMetrics would be one: we can ignore the randomization of decimal points with a massive font size

**Comment by Thorin-Oakenpants** [comment]
Sweet: I added fontPixel and fontSize

What would you call them? currently TZP labels are `[offset] fonts` `[client] fonts` and `[scroll] fonts`. `pixel` and `pixel size` seem ambiguous / too similar

What is your lie logic and return now?

**Comment by abrahamjuliot** [comment]
I use `pixel` and `size`, but technically, all the css tests are just different ways to get the computed pixel dimensions, so you could call them `pixel 1`, `pixel 2`, etc. 

It's not noted above, but `inlineSize` and `blockSize` are controlled by `writing-mode`, so I've added `writing-mode: horizontal-tb !important;` to the span css.

I also added the `perspective-origin` and `transform-origin` tests. I think that's it for the pixel category.

```css
/* add this to the span css for accurate origin calculations*/
transform-origin: unset !important;
perspective-origin: unset !important;
```
```js
//....

const pixelsToInt = pixels => Math.round(+pixels.replace('px',''))
const originPixelsToInt = pixels => Math.round(2*pixels.replace('px', ''))
const style = getComputedStyle(span)

const getDimensions = (span, style) => {
    const transform = style.transformOrigin.split(' ')
    const perspective = style.perspectiveOrigin.split(' ')
    const dimensions = {
        width: pixelsToInt(style.width),
        height: pixelsToInt(style.height),
        transformWidth: originPixelsToInt(transform[0]),
        transformHeight: originPixelsToInt(transform[1]),
        perspectiveWidth: originPixelsToInt(perspective[0]),
        perspectiveHeight: originPixelsToInt(perspective[1]),
        sizeWidth: pixelsToInt(style.inlineSize),
        sizeHeight: pixelsToInt(style.blockSize),
        scrollWidth: span.scrollWidth,
        scrollHeight: span.scrollHeight,
        offsetWidth: span.offsetWidth,
        offsetHeight: span.offsetHeight,
        clientWidth: span.clientWidth,
        clientHeight: span.clientHeight
    }
    return dimensions
}

// evaluate this in the loop
const dimensions = getDimensions(span, style)
```

**Comment by Thorin-Oakenpants** [comment]
thanks, already had all that .. now I see two more, transform and perspective

**Comment by Thorin-Oakenpants** [comment]
@abrahamjuliot (who is always busy)

OK, why do I get the wrong results for perspective and transform? It's returning them all (486/486)
![fonts](https://user-images.githubusercontent.com/16656956/104038683-54ee9700-51cd-11eb-99c7-e3b04152bbe6.png)

https://github.com/abrahamjuliot/creepjs/blob/0c70c13c10fbbb501375034a4fc0705b436f1808/modules/fonts.js

lie detector
```js
const createLieDetector = () => {
	/* https://github.com/abrahamjuliot/creepjs */
	let invalidDimensions = []
	return {
		getInvalidDimensions: () => invalidDimensions,
			compute: ({
				width,
				height,
				transformWidth,
				transformHeight,
				perspectiveWidth,
				perspectiveHeight,
				sizeWidth,
				sizeHeight,
				scrollWidth,
				scrollHeight,
				offsetWidth,
				offsetHeight,
				clientWidth,
				clientHeight
		}) => {
			const invalid = (
				width != transformWidth ||
				width != perspectiveWidth ||
				width != sizeWidth ||
				width != scrollWidth ||
				width != offsetWidth ||
				width != clientWidth ||

				height != transformHeight ||
				height != perspectiveHeight ||
				height != sizeHeight ||
				height != scrollHeight ||
				height != offsetHeight ||
				height != clientHeight
			)
			if (invalid) {
				invalidDimensions.push({
					width: [width, transformWidth, perspectiveWidth, sizeWidth, scrollWidth, offsetWidth, clientWidth],
					height: [height, transformHeight, perspectiveHeight, sizeHeight, scrollHeight, offsetHeight, clientHeight]
				})
			}
			return
		}
	}
}
```

getFonts
```js
const getFonts = () => {
	/* https://github.com/abrahamjuliot/creepjs */
	return new Promise(resolve => {
		if (!isFF) {
			return resolve("n/a")
		}
		try {
			const detectLies = createLieDetector()
			const doc = document // or iframe.contentWindow.document
			const id = `font-fingerprint`
			const div = doc.createElement('div')
			div.setAttribute('id', id)
			doc.body.appendChild(div)
			doc.getElementById(id).innerHTML = `
				<style>
				#${id}-detector {
					--font: '';
					position: absolute !important;
					left: -9999px!important;
					font-size: 256px !important;
					font-style: normal !important;
					font-weight: normal !important;
					letter-spacing: normal !important;
					line-break: auto !important;
					line-height: normal !important;
					text-transform: none !important;
					text-align: left !important;
					text-decoration: none !important;
					text-shadow: none !important;
					white-space: normal !important;
					word-break: normal !important;
					word-spacing: normal !important;
					/* in order to test scrollWidth, clientWidth, etc. */
					padding: 0 !important;
					margin: 0 !important;
					/* in order to test inlineSize and blockSize */
					writing-mode: horizontal-tb !important;

					transform-origin: unset !important;
					perspective-origin: unset !important;

				}
				#${id}-detector::after {
					font-family: var(--font);
					content: '` + fntStrA + `';
				}
				</style>
				<span id="${id}-detector"></span>`

			const span = doc.getElementById(`${id}-detector`)
			const pixelsToInt = pixels => Math.round(+pixels.replace('px',''))
			const originPixelsToInt = pixels => Math.round(2*pixels.replace('px', ''))
			const detectedViaPixel = new Set()
			const detectedViaPixelSize = new Set()
			const detectedViaScroll = new Set()
			const detectedViaOffset = new Set()
			const detectedViaClient = new Set()
			const detectedViaTransform = new Set()
			const detectedViaPerspective = new Set()
			const baseFonts = ['monospace', 'sans-serif', 'serif']
			const style = getComputedStyle(span)

			const getDimensions = (span, style) => {
				const transform = style.transformOrigin.split(' ')
				const perspective = style.perspectiveOrigin.split(' ')
				const dimensions = {
					width: pixelsToInt(style.width),
					height: pixelsToInt(style.height),
					transformWidth: originPixelsToInt(transform[0]),
					transformHeight: originPixelsToInt(transform[1]),
					perspectiveWidth: originPixelsToInt(perspective[0]),
					perspectiveHeight: originPixelsToInt(perspective[1]),
					sizeWidth: pixelsToInt(style.inlineSize),
					sizeHeight: pixelsToInt(style.blockSize),
					scrollWidth: span.scrollWidth,
					scrollHeight: span.scrollHeight,
					offsetWidth: span.offsetWidth,
					offsetHeight: span.offsetHeight,
					clientWidth: span.clientWidth,
					clientHeight: span.clientHeight
				}
				return dimensions
			}
			const base = baseFonts.reduce((acc, font) => {
				span.style.setProperty('--font', font)
				const dimensions = getDimensions(span, style)
				detectLies.compute(dimensions)
				acc[font] = dimensions
				return acc
			}, {})

			const families = fntList.reduce((acc, font) => {
				baseFonts.forEach(baseFont => acc.push(`'${font}', ${baseFont}`))
				return acc
			}, [])

			families.forEach(family => {
				span.style.setProperty('--font', family)
				const basefont = /, (.+)/.exec(family)[1]
				const style = getComputedStyle(span)
				const dimensions = {
					width: pixelsToInt(style.width),
					height: pixelsToInt(style.height),
					sizeWidth: pixelsToInt(style.inlineSize),
					sizeHeight: pixelsToInt(style.blockSize),
					scrollWidth: span.scrollWidth,
					scrollHeight: span.scrollHeight,
					offsetWidth: span.offsetWidth,
					offsetHeight: span.offsetHeight,
					clientWidth: span.clientWidth,
					clientHeight: span.clientHeight
				}
				detectLies.compute(dimensions)
				const font = /\'(.+)\'/.exec(family)[1]
				if (dimensions.width != base[basefont].width ||
					dimensions.height != base[basefont].height) {
					detectedViaPixel.add(font)
				}
				if (dimensions.sizeWidth != base[basefont].sizeWidth ||
					dimensions.sizeHeight != base[basefont].sizeHeight) {
					detectedViaPixelSize.add(font)
				}
				if (dimensions.scrollWidth != base[basefont].scrollWidth ||
					dimensions.scrollHeight != base[basefont].scrollHeight) {
					detectedViaScroll.add(font)
				}
				if (dimensions.offsetWidth != base[basefont].offsetWidth ||
					dimensions.offsetHeight != base[basefont].offsetHeight) {
					detectedViaOffset.add(font)
				}
				if (dimensions.clientWidth != base[basefont].clientWidth ||
					dimensions.clientHeight != base[basefont].clientHeight) {
					detectedViaClient.add(font)
				}
				if (dimensions.transformWidth != base[basefont].transformWidth ||
					dimensions.transformHeight != base[basefont].transformHeight) {
					detectedViaTransform.add(font)
				}
				if (dimensions.perspectiveWidth != base[basefont].perspectiveWidth ||
					dimensions.perspectiveHeight != base[basefont].perspectiveHeight) {
					detectedViaPerspective.add(font)
				}
				return
			})
			const fontsPixel = [...detectedViaPixel]
			const fontsPixelSize = [...detectedViaPixelSize]
			const fontsScroll = [...detectedViaScroll]
			const fontsOffset = [...detectedViaOffset]
			const fontsClient = [...detectedViaClient]
			const fontsPerspective = [...detectedViaPerspective]
			const fontsTransform = [...detectedViaTransform]
			return resolve({
				lies: !!detectLies.getInvalidDimensions().length,
				fontsScroll,
				fontsOffset,
				fontsClient,
				fontsPixel,
				fontsPixelSize,
				fontsPerspective,
				fontsTransform
			})
		} catch (error) {
			console.error(error)
			return resolve("error")
		}
	})
}
```

get_fonts()
```js
function get_fonts() {
	return new Promise(resolve => {
		let t0 = performance.now(),
			fontReturn = []

		function finish_up() {
			// snip
			return resolve(fontReturn)
		}

		// run it
		getFonts().then(res => {
			// remove element
			try {document.getElementById("font-fingerprint").remove()} catch(e) {}

			// handled other cases
			if (res == "n/a") {
				fontReturn = ["fonts_hash:n/a","fonts_count:n/a","fonts_lied:n/a"]
				finish_up()
			} else if (res == "error") {
				fontReturn = ["fonts_hash:error","fonts_count:error","fonts_lied:error"]
				//snip
				finish_up()
			} else {
				const {lies, fontsScroll, fontsOffset, fontsClient, fontsPixel, fontsPixelSize, fontsPerspective, fontsTransform} = res
				let isLies = lies
				// values
				let hashO = sha1(fontsOffset.join()),
					countO = fontsOffset.length
				let hashC = sha1(fontsClient.join()),
					countC = fontsClient.length
				let hashS = sha1(fontsScroll.join()),
					countS = fontsScroll.length
				let hashP = sha1(fontsPixel.join()),
					countP = fontsPixel.length
				let hashPS = sha1(fontsPixelSize.join()),
					countPS = fontsPixelSize.length
				let hashPR = sha1(fontsPerspective.join()),
					countPR = fontsPerspective.length
				let hashT = sha1(fontsTransform.join()),
					countT = fontsTransform.length
				let countF = fntList.length

				// output
				dom.fontOffset.innerHTML = hashO + s12 + "["+countO+"/"+countF+"]" + sc
				dom.fontClient.innerHTML = hashC + s12 + "["+countC+"/"+countF+"]" + sc
				dom.fontScroll.innerHTML = hashS + s12 + "["+countS+"/"+countF+"]" + sc
				dom.fontPixel.innerHTML = hashP + s12 + "["+countP+"/"+countF+"]" + sc
				dom.fontPixelSize.innerHTML = hashPS + s12 + "["+countPS+"/"+countF+"]" + sc
				dom.fontPerspective.innerHTML = hashPR + s12 + "["+countPR+"/"+countF+"]" + sc
				dom.fontTransform.innerHTML = hashT + s12 + "["+countT+"/"+countF+"]" + sc

				// snip: set fontReturn
				finish_up()
			}
		})
	})
}
```

**Comment by abrahamjuliot** [comment]
In `families.forEach(family => { })`, replace `const dimensions = { width: pixelsToInt(style.width)...` with `const dimensions = getDimensions(span, style)`.

**Comment by Thorin-Oakenpants** [comment]
^^ done: seems to work. I already have that line: I've left it in, I assume it's needed?

```js
			const base = baseFonts.reduce((acc, font) => {
				span.style.setProperty('--font', font)
				const dimensions = getDimensions(span, style) // <--- HERE
				detectLies.compute(dimensions)
				acc[font] = dimensions
				return acc
			}, {})
```

**Comment by abrahamjuliot** [comment]
Looks good. Yes, `getDimensions` will return the goods for both the base and family fonts.

**Comment by Thorin-Oakenpants** [comment]
thanks... I want to expand on these fonts tests (in a new issue/ToDo item), as I mentioned earlier with textmetrics and svg etc - which you already added to your font test page. Just not sure how deep/much I want to do in TZP itself

BTW I get a non-matching counts on mine (vanilla profile FF nightly)
![notmatch](https://user-images.githubusercontent.com/16656956/104088623-279ef900-5260-11eb-839d-d9a6d8f55175.png)

Is SVG and TM produce legit different results, then that might be more entropy .. or it could be equivalency .. who knows!!

I'll leave this open, for this issue, as there are false positives in older FF's: namely `size` test returns 0 fonts found: maybe that needs a check to see if the API is supported

**Comment by Thorin-Oakenpants** [comment]
@abrahamjuliot .. whats the easiest way to count occurrences of items

e.g. I have an array of 7 hashes `[a,b,a,b,a,a,a]` and I just want the most common: i.e answer = `item a, count = 4`
- doesn't matter if there are two most common: e.g. `[a,a,a,b,b,b,c]` .. i.e A can be `a or b` and `count = 3`

**Comment by abrahamjuliot** [comment]
> non-matching counts

Same here. SVG and TextMetrics are non-matching in both Chrome and FF. In FF they are distinctly unique. I'm not sure why that is, but it appears to be unique rendering precisions in 2d canvas, vector graphics, and html element text.

serif font

- TextMetrics
actualBoundingBoxAscent: 178
width: 2205

- DOMRect
clientHeight: 300
clientWidth: 2205

- SVG
height: 280
width: 2204

> size

[blockSize](https://developer.mozilla.org/en-US/docs/Web/CSS/block-size) and [inlineSize](https://developer.mozilla.org/en-US/docs/Web/CSS/inline-size) support can be checked with `'blockSize' in document.body.style`.

> count occurrences of items

I gained inspiration from [this](https://stackoverflow.com/a/1053865) and used [Array.reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce). This modification uses `JSON.stringify` to cover an array of object and/or arrays.

```js
const greatest = list.reduce((greatest, currentValue) => {
	const stringValue = JSON.stringify(currentValue)
	if (!greatest.elements[stringValue]) {
		greatest.elements[stringValue] = 1
	}
	else {
		greatest.elements[stringValue]++
	}
	if (greatest.elements[stringValue] > greatest.count) {
		greatest.count = greatest.elements[stringValue]
		greatest.item = currentValue
	}
	return greatest
}, { elements: {}, count: 1, item: list[0] })

greatest.count // => 2
greatest.item // => {a: 5}
greatest.elements /* => 
{
  {"a":0}: 1,
  {"a":2}: 1,
  {"a":4}: 1,
  {"a":5}: 2,
  {"b":1}: 1,
  {"b":3}: 1
}
*/
````

Another method is to use `reduce` and `filter`.

```js
const list = [{a: 0},{b: 1},{a: 2},{b: 3},{a: 4},{a: 5},{a: 5}]
const greatest = list.reduce((greatest , currentValue, index, list) => {
    const count = list.filter(item => JSON.stringify(item) == JSON.stringify(currentValue)).length
    if (count > greatest.count) {
        return { count, item: currentValue }
    }
    return greatest
}, { count: 0, item: undefined })

greatest.count // => 2
greatest.item // => {a: 5}
```

**Comment by Thorin-Oakenpants** [comment]
```js
const list = [{a: 0},{b: 1},{a: 2},{b: 3},{a: 4},{a: 5},{a: 5}]
```

what is 0 thru to 5? Why two 5's? If that's the index position, then that seems redundant

**Comment by abrahamjuliot** [comment]
The object/props are unrelated. `{a: 5}` has a duplicate to show objects can be counted, but the list can contain anything.

**Comment by Thorin-Oakenpants** [comment]
ok, so that's not returning what I need .. in an array of seven items which contains 5 a's and 2 b's then greatest count is not 2 (unique items), it's 5 (the number of a's)

**Comment by abrahamjuliot** [comment]
My list is just an arbitrary list as an example input. `getGreatestOccurrence` can take `[a, a, a, a, a, b, b]` as the argument and it will return count 5.

```js
const a = 's6%2yjTz&yR'
const b = '@^xpU2yHovj'
const list = [a, a, a, a, a, b, b,]

const greatest = getGreatestOccurrence(list)
greatest.count // => 5
greatest.item // => "s6%2yjTz&yR"
```

**Comment by Thorin-Oakenpants** [comment]
^^ cool. Sorry, the `greatest.count // => 2` in the examples confuses me

**Comment by Thorin-Oakenpants** [comment]
Imma gonna close this: I do want to add domrect, svg, textmetric font tests, and that is listed in my ToDo issue: but I'll open a new issue if/when I get to it. Domrect is equivalency. Textmetic and svg is interesting for the different values = maybe some extra entropy

PS: I went with if four or more of the seven are the same then that's legit otherwise I can't tell. Thanks for all that @abrahamjuliot .. here, have some pudding 🍰
